경북대 BE 박신영 3단계 - 상품 옵션

parknew0 opened the pull request:
안녕하세요 멘토님! 3단계 구현을 완료하였습니다.

## 주요 변경 내용
### 상품-옵션 연관관계 매핑: 
Product와 Option을 1:N 관계로 설정하고, 상품 저장/수정/삭제 시 옵션도 함께 관리되도록 Cascade와 orphanRemoval을 적용했습니다.

### 상품 및 옵션 동시 생성/수정: 
상품을 등록하거나 수정할 때, 옵션 목록까지 한 번의 API 요청으로 처리할 수 있도록 ProductService 로직을 구현했습니다.

### 옵션 목록 조회 API 추가: 
특정 상품에 속한 모든 옵션을 조회하는 GET /api/products/{productId}/options API를 추가했습니다.

### 유효성 검사: 
상품 내에서 옵션 이름이 중복되지 않도록 서버 단에서 검증하는 로직을 추가했습니다.

parknew0 committed:
feat(domain): Option 엔티티 클래스 생성

- id, name, quantity 필드와 각 필드에 대한 제약조건 포함
- 이름, 수량에 대한 유효성 검사 어노테이션 적용

parknew0 committed:
feat(domain): Product와 Option 엔티티 연관관계 매핑

Product와 Option 간의 1:N 양방향 연관관계를 설정

parknew0 committed:
feat(repository): OptionRepository 생성

parknew0 committed:
feat(domain): Option 엔티티에 수량 차감(subtract) 로직 추가

parknew0 committed:
feat(dto): Option 관련 DTO 생성 및 ProductRequest 수정

- `OptionRequest`, `OptionResponse` 레코드 새로 생성
- `ProductRequest`에 `List<OptionRequest>`를 추가

parknew0 committed:
feat(service): ProductService에 상품 및 옵션 추가 기능 구현

parknew0 committed:
feat(service): OptionService 생성 및 옵션 조회 기능 추가

parknew0 committed:
feat(controller): 특정 상품의 옵션 목록 조회 API 구현

parknew0 committed:
feat(service): 상품 및 옵션 수정 기능 구현

parknew0 committed:
feat(test): 옵션 기능 관련 단위 테스트 및 통합 테스트 작성

leaf-upper commented on src/main/java/gift/entity/Product.java:
```diff
@@ -1,26 +1,29 @@
 @Table(name = "product")
 public class Product {
 
+    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
```
cascade = CascadeType.ALL, orphanRemoval = true
해당 2가지 옵션에 대해서 간략하게 설명해주시면 좋을것 같아요.

leaf-upper replied:
Product를 가져올때 OneToMany에 연관되어있는 Option을 전부 조회해온다면, 굳이 OptionRepository가 필요할까요?

parknew0 replied:
cascade = CascadeType.ALL는 부모 엔티티의 상태 변화(저장, 수정, 삭제 등)를 자식 엔티티에게도 그대로 전달하는 설정입니다.
orphanRemoval = true는부모 엔티티와의 연관 관계가 끊어진 자식 엔티티를 고아로 간주하고 데이터베이스에서 자동으로 삭제하는 기능입니다!

parknew0 replied:
OptionRepository는 추후 옵션 관련 기능 확장을 염두에 두고 만들어뒀습니다! product에 딸린 option들을 조회할 때 OptionRepository를 사용하는 것을 고려해봤지만, 이미 조회된 product 객체에게 Option들을 달라고 하는 편이 더 자연스러운 것 같아서 OptionRepository를 지금 단계에선 사용하지 않았습니다!
특정 시작 객체가 존재하지 않을 때와 같은 경우에 OptionRepository를 사용할 것을 생각해 봤습니다.

leaf-upper commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -0,0 +1,30 @@
+        this.productRepository = productRepository;
+    }
+
+    public List<OptionResponse> getOptionsForProduct(Long productId) {
```
DB 테이블 단위로 Repository, Service를 만들어가는건 저는 별로 추천드리지 않습니다.
ProductService에서 Product를 조회할때 연관되어있는 Option을 전부 같이 조회할수있고, 관련된 로직도 ProductService와 Product에서 할 수 있을거라고 생각이 들어요
한번 더 고민해보면 좋을것 같습니다.

parknew0 replied:
Product와 Option은 긴밀하게 연결되어있는 객체들이고, Option은 Product에 종속된 기능이니
ProductService가 서비스 책임을 가져가는 것이 자연스럽다는 말씀이시네요!

더 찾아보니 애그리거트라는 개념이 있는데 정리해보자면 아래의 내용인 것 같아요.
```
외부에서는 루트만 바라본다: 
다른 서비스나 컨트롤러는 애그리거트 내부의 Option에 직접 접근하는 것이 아니라, 
오직 루트인 Product에만 접근해야 한다.

모든 변경은 루트를 통해 시작된다: 
Option을 추가하거나 삭제하는 로직이 필요하다면,
Option을 직접 수정하는 것이 아니라 Product 객체에게 요청해야 한다.

하나의 트랜잭션으로 관리된다: 
애그리거트의 상태를 변경할 때는 루트를 저장(productRepository.save(product))하는 것만으로
묶음 전체의 변경사항이 하나의 트랜잭션 안에서 처리된다.
```

좋은 인사이트 감사합니다!

leaf-upper replied:
@parknew0 
네 사실 애그리거트라는 용어를 쓰지 않고 설명을 드리려고 했는데, 인터넷에서 직접 찾으셨다니 대단합니다.
애그리거트는 도메인 주도 개발에서 주로 나오는 용어인데요, 시간이 되실때 도메인 주도 개발에서 Repository의 역할에 대해서도
한번 확인해보시면 좋을것 같습니다. (Jpa에서 사용하는 Repository와 같은 의미입니다.)

leaf-upper reviewed (changes requested):
안녕하세요 신영님
Option 추가해주신 부분 잘 읽어봤습니다.
몇가지 코멘트를 남겨드렸는데요, 확인해보시고 
코드 수정후에 다시 요청주시면 리뷰드리겠습니다.
고생하셨습니다!

parknew0 committed:
refactor(product): Product 서비스에 애그리거트 패턴 적용

- `OptionService` 삭제 및 `ProductService`로 기능 통합
- `ProductController`의 의존성을 `ProductService`로 단일화

parknew0 committed:
refactor(product): Product 엔티티의 캡슐화 강화

Product 애그리거트의 데이터 일관성과 관련된 책임과 로직을 Product 엔티티 내부로 이전하여 캡슐화를 강화했습니다.

leaf-upper reviewed (approved):
👍 👍 고생하셨습니다. 신영님
제가 리뷰드린부분에 대해서 잘 수정해주셨고
다음 과제 진행해주셔도 좋을것 같습니다.
고생많으셨습니다 어프로브 드릴게요!
